### ISSUE TITLE
NodePattern fails extracting node types when presented with nil

### ISSUE THREAD

**Issue by shalecraig** [rubocop-hq/rubocop#5470]
### Ruby Code:
A short test case for this problem:
```ruby
pattern = RuboCop::NodePattern.new('(send ($send ...) _method)')
source = 'a' # "(send nil :a)"
node = RuboCop::ProcessedSource.new(source, RUBY_VERSION.to_f, nil).ast
type = pattern.match(node)
```
#### Expected behavior

`pattern.match(node)` returns `nil`

#### Actual behavior

`pattern.match(node)` raises:

```
NoMethodError: undefined method `type' for nil:NilClass
```

### Test variant that works:
```ruby
pattern = RuboCop::NodePattern.new('(send $(send ...) _method)')
source = 'a' # "(send nil :a)"
node = RuboCop::ProcessedSource.new(source, RUBY_VERSION.to_f, nil).ast
result = pattern.match(node)
type = result&.type
```

#### Why this is relevant:

I'm trying to write a `NodePattern` that captures the difference between a `send` and `csend`.

#### Workaround:
For now, I can avoid this problem by:
* Specifying `RuboCop::NodePattern.new('(send $({send csend} ...) _method)')` instead of `RuboCop::NodePattern.new('(send (${send csend} ...) _method)')`
* Extracting the node's type based on `result.type`

## RuboCop version

```
$ bundle exec rubocop -V
0.49.1 (using Parser 2.4.0.2, running on ruby 2.4.1 x86_64-darwin16)
```

**Comment by Drenmi** [comment]
Labelling this as a feature request, as I don't think this is something that was intended to work. 🙂

**Comment by thomthom** [comment]
I'm running into similar error:

The custom cop I'm working on;

```ruby
        def_node_matcher :sketchup_require, <<-PATTERN
          (send
            (const nil? :Sketchup) :require
            (str $_)
          )
        PATTERN

        def_node_matcher :sketchup_require?, <<-PATTERN
          (send
            (const nil? :Sketchup) :require
            (str _)
          )
        PATTERN


        def_node_matcher :sketchup_extension_new, <<-PATTERN
          (:send
            (:const nil? :SketchupExtension) :new
            _
            (str $_))
        PATTERN

        def_node_search :sketchup_extension_new?, <<-PATTERN
          (:send
            (:const nil? :SketchupExtension) :new
            _
            (str _))
        PATTERN


        def on_send(node)
          if sketchup_require?(node)
            filename = sketchup_require(node)
          elsif sketchup_extension_new?(node)
            filename = sketchup_extension_new(node)
          else
            return
          end
          add_offense(node, location: :expression) unless valid_filename?(filename)
        end
```

Here are the tests that pass (as expected);

```
  it 'registers an offense when using `Sketchup.require` with file extension' do
    inspect_source('Sketchup.require("filename.rb")')
    expect(cop.offenses.size).to eq(1)
  end

  context 'SketchupExtension.new' do

    it 'registers an offense when using `SketchupExtension.new` with file extension' do
      inspect_source('extension = SketchupExtension.new("Example", "Example/main.rb")')
      expect(cop.offenses.size).to eq(1)
    end

  end
```

However, this tests fail with an error;

```ruby
  it 'does not register an offense when using `require` without file extension' do
    inspect_source('require("filename")')
    expect(cop.offenses).to be_empty
  end
```

```
  1) RuboCop::Cop::SketchupSuggestions::SketchupRequire does not register an offense when using `require` without file extension
     Failure/Error: elsif sketchup_extension_new?(node)

     NoMethodError:
       undefined method `type' for nil:NilClass
     # C:3:in `block in sketchup_extension_new?'
     # C:2:in `sketchup_extension_new?'
     # ./lib/rubocop/sketchup/suggestions/sketchup_require.rb:47:in `on_send'
     # ./spec/rubocop/sketchup/suggestions/sketchup_require_spec.rb:38:in `block (2 levels) in <top (required)>'
```

It seems odd to me if that's expected to raise an error. If it is, what would be recommended to check for these various method signatures I'm trying to match?

```
$ bundle exec rubocop -V
0.53.0 (using Parser 2.5.0.3, running on ruby 2.4.3 i386-mingw32)
```

**Comment by thomthom** [comment]
Hm.. re-reading the original post closer I'm wondering if my issue isn't exactly the same. I'm not trying to match type, but method signature...

**Comment by marcandre** [comment]
I'm on it...

**Comment by marcandre** [comment]
Bug summary: the matcher would raise an error if most types of node matching / capturing / ... is attempted on a literal instead of an `RuboCop::AST::Node`.
